### PR TITLE
adding new relation, reverse and aggregation

### PR DESCRIPTION
**Pull request by nightfury394**

### CODE REVIEW

File: project/Admin.cs [hunk 1]
```diff
@@ -56,17 +63,36 @@ TimeSpan endTime
...
+            var timeslot = ManagedTimeslots.Find(t => t.ScheduleId == scheduleId);
+            if (timeslot != null)
+            {
+                timeslot.Date = date;
+                timeslot.UpdateTime(startTime, endTime);
+                Console.WriteLine(
+                    $"Schedule ID {scheduleId} updated by Admin {Name}: {date.ToShortDateString()}, {startTime} - {endTime}"
+                );
+            }
+            else
```
Comment (hussalen) [inline project/Admin.cs]: Else unnecessary

File: project/Admin.cs [hunk 2]
```diff
@@ -80,5 +106,17 @@ public Report GenerateReport(string title, JsonArray content)
             );
             return report;
         }
+
+        public void AddTimeslot(Timeslot timeslot)
+        {
+            if (timeslot == null)
+                throw new ArgumentException("Timeslot cannot be null.");
+
+            if (!ManagedTimeslots.Contains(timeslot))
```
Comment (hussalen) [inline project/Admin.cs]: And what if the timeslot is contained in the data structure? (we need to throw an exception if that is the case here)

File: project/Admin.cs [hunk 3]
```diff
@@ -56,17 +63,36 @@ TimeSpan endTime
...
                 throw new ArgumentException("Schedule ID must be a positive integer.");
 
-            Console.WriteLine($"Schedule ID {scheduleId} deleted by Admin {Name}");
+            var timeslot = ManagedTimeslots.Find(t => t.ScheduleId == scheduleId);
+            if (timeslot != null)
+            {
+                ManagedTimeslots.Remove(timeslot);
+                Console.WriteLine($"Schedule ID {scheduleId} deleted by Admin {Name}");
+            }
+            else
```
Comment (hussalen) [inline project/Admin.cs]: else not needed

File: project/Admin.cs [hunk 4]
```diff
@@ -21,6 +23,8 @@ public string Name
             }
         }
 
+        public List<Timeslot> ManagedTimeslots { get; private set; } = new();
```
Comment (hussalen) [inline project/Admin.cs]: If a getter is called, we need to return a copy of the List (applies to all reverse connections that involve a data structure of some kind)...

File: project/Admin.cs (hunk 1 above)
Comment (nightfury394) [reply project/Admin.cs]: Okay

File: project/Admin.cs (hunk 2 above)
Comment (nightfury394) [reply project/Admin.cs]: will add logic to explicitly throw an exception if the timeslot already exists in the ManagedTimeslots list .

File: project/Admin.cs (hunk 3 above)
Comment (nightfury394) [reply project/Admin.cs]: okay

File: project/Admin.cs (hunk 4 above)
Comment (nightfury394) [reply project/Admin.cs]: will update the getter to return a new list containing the current elements, ensuring encapsulation and immutability

File: project/Assignment.cs [hunk 5]
```diff
@@ -2,19 +2,76 @@
...
-        public static int nextId;
+        public static int nextId; 
         public string Topic { get; set; }
-
         public DateTime DueDate { get; set; }
         public DateTime? SubmissionDate { get; set; }
+        public Student SubmittingStudent { get; set; }
+    }
 
+    public class Assignment : IAssignment
```
Comment (hussalen) [inline project/Assignment.cs]: Why are you creating a concrete class called Assignment if it's supposed to be represented as an interface? The inheritors are IndividualProject, GroupProject, Essay and Coding.

**Review by hussalen — CHANGES_REQUESTED** [review]
Please address every comment that I've added, then we can approve this for merging

@nightfury394

File: project/Assignment.cs (hunk 5 above)
Comment (nightfury394) [reply project/Assignment.cs]: sorry my bad . i thought its different . u are right

File: project/Assignment.cs [hunk 6]
```diff
@@ -5,29 +5,118 @@
...
-        public static int nextId; 
+        int AssignmentID { get; }
+        string Topic { get; set; }
+        DateTime DueDate { get; set; }
+        DateTime? SubmissionDate { get; set; }
+        Student SubmittingStudent { get; set; }
+    }
+
+   
+    public class IndividualProject : IAssignment
```
Comment (hussalen) [inline project/Assignment.cs:19]: Dude, we have those in separate classes...

Comment (nightfury394) [reply project/Assignment.cs:19]: but we dont put them in interface ?

File: project/Admin.cs [hunk 7]
```diff
@@ -23,7 +23,13 @@ public string Name
             }
         }
 
-        public List<Timeslot> ManagedTimeslots { get; private set; } = new();
+        
+        public List<Timeslot> ManagedTimeslots
```
Comment (hussalen) [inline project/Admin.cs:27]: How are we going to set or remove the time slot from admin?

**Pull request closed without merging**